WooCommerce/Classes/ViewRelated/Editor/AztecEditorViewController.swift

```diff
@@ -0,0 +1,216 @@
+                  textViewAttachmentDelegate: TextViewAttachmentDelegate = AztecTextViewAttachmentHandler()) {
+        self.content = content ?? ""
+        self.textViewAttachmentDelegate = textViewAttachmentDelegate
+        self.viewProperties = viewProperties
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        registerAttachmentImageProviders()
+
+        configureNavigationBar()
+        configureView()
+        configureSubviews()
+
+        aztecUIConfigurator.configureConstraints(editorView: editorView,
+                                                 editorContainerView: view,
+                                                 placeholderView: placeholderLabel)
+
+        setHTML(content)
+
+        refreshPlaceholderVisibility()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        startListeningToNotifications()
+    }
+}
+
+private extension AztecEditorViewController {
+    func configureNavigationBar() {
+        title = viewProperties.navigationTitle
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(saveButtonTapped))
+    }
+
+    func configureView() {
+        edgesForExtendedLayout = UIRectEdge()
+        view.backgroundColor = StyleManager.wooWhite
+    }
+
+    func configureSubviews() {
+        view.addSubview(richTextView)
+        view.addSubview(htmlTextView)
+        view.addSubview(placeholderLabel)
+    }
+
+    func registerAttachmentImageProviders() {
+        let providers: [TextViewAttachmentImageProvider] = [
+            SpecialTagAttachmentRenderer(),
+            CommentAttachmentRenderer(font: StyleManager.subheadlineBoldFont),
+            HTMLAttachmentRenderer(font: StyleManager.subheadlineBoldFont),
+            GutenpackAttachmentRenderer()
+        ]
+
+        for provider in providers {
+            richTextView.registerAttachmentImageProvider(provider)
+        }
+    }
+}
+
+private extension AztecEditorViewController {
+    func setHTML(_ html: String) {
+        editorView.setHTML(html)
+    }
+
+    func getHTML() -> String {
+        return editorView.getHTML()
+    }
+
+    func refreshPlaceholderVisibility() {
+        placeholderLabel.isHidden = richTextView.isHidden || !richTextView.text.isEmpty
+    }
+}
+
+// MARK: Keyboard frame update handling
+//
+private extension AztecEditorViewController {
+    func handleKeyboardFrameUpdate(keyboardFrame: CGRect) {
+        let referenceView = editorView.activeView
+
+        let contentInsets  = UIEdgeInsets(top: referenceView.contentInset.top,
+                                          left: 0,
+                                          bottom: view.frame.maxY - (keyboardFrame.minY + self.view.layoutMargins.bottom),
```

Comment:
Not sure it considers the safe area and the tab bar.
In fact when I open the keyboard and I have a very long text when I close the keyboard the text area ends up under the tab bar.

Reply:
great catch! I spent some time looking into this and was hoping to not do this manual bottom inset adjustment by setting the bottom inset or the view controller's bottom inset `additionalSafeAreaInsets.bottom` with the keyboard height. However, I was seeing a weird gap between the keyboard and the editor text view's bottom inset on both the simulators and a device.

Eventually, I got the bottom inset at the right place by preventing a negative bottom inset here: https://github.com/woocommerce/woocommerce-ios/pull/1433/commits/35bf9f3eba6b551c5d64e5e5f9f5761fa6938698 - see if it works for you now!

I think it wasn't an issue before (like in the manual tracking where we adjusted the scroll view's bottom inset with the keyboard height) because the view is a fullscreen modal and doesn't have to consider the tab bar. Let me know if you have better solutions for the bottom inset + keyboard height + tab bar!